torchsynth/synth.py

```diff
@@ -55,30 +55,32 @@ def buffer_size(self) -> T:
         assert self.synthglobals.buffer_size.ndim == 0
         return self.synthglobals.buffer_size
 
-    def add_synth_modules(self, modules: List[Tuple[str, SynthModule]]):
+    def add_synth_modules(
+        self, modules: List[Tuple[str, SynthModule, Optional[Dict[str, Any]]]]
+    ):
         """
         Add a set of named children TorchSynthModules to this synth. Registers them
         with the torch nn.Module so that all parameters are recognized.
 
         Args:
-            modules List[Tuple[str, SynthModule]]: A list of SynthModules and
-                                            their names.
+            modules List[Tuple[str, SynthModule, Optional[Dict[str, Any]]]]: A list of
+                SynthModule classes with their names and any parameters to pass to
+                their constructor.
         """
 
-        for name, module in modules:
-            if not isinstance(module, SynthModule):
-                raise TypeError(f"{module} is not a SynthModule")
-
-            if module.batch_size != self.batch_size:
-                raise ValueError(f"{module} batch_size does not match")
-
-            if module.sample_rate != self.sample_rate:
-                raise ValueError(f"{module} sample rate does not match")
+        for module_tuple in modules:
+            if len(module_tuple) == 3:
+                name, module, params = module_tuple
+            else:
+                name, module = module_tuple
+                params = {}
 
-            if module.buffer_size != self.buffer_size:
-                raise ValueError(f"{module} buffer size does not match")
+            if not issubclass(module, SynthModule):
+                raise TypeError(f"{module} is not a SynthModule")
 
-            self.add_module(name, module)
+            self.add_module(
+                name, module(self.synthglobals, device=self.device, **params)
+            )
```

Comment:
This is a cool pattern, good idea @turian  -- now if any modules are added to the Synth after construction then they constructed on the device. We don't have anything like that now, but could allow for some interesting modular architectures

Reply:
Also nice b/c we don't have to have all those checks on synthglobals